### ISSUE TITLE
Grid does not reduce height request.

### ISSUE THREAD

**Issue by KonyukhovSergey** [dotnet/maui#7298]
### Description

Grid does not layout own children correctly. See the image below. Left is expected layout, right is as in MAUI.
<img width="717" alt="image" src="https://user-images.githubusercontent.com/1398566/169093753-4f6f1607-8dc9-406a-9904-9a89adf01845.png">

Also, in the landscape orientation, when the content page does not have enough vertical space the grid does not reduce child height. The first image is expected, the second is in fact.
<img width="565" alt="image" src="https://user-images.githubusercontent.com/1398566/169097002-e5427667-6334-4136-a161-f8aaffd7a89b.png">

### Steps to Reproduce

Create new MAUI app.
Use the following XAML in the MainPage.xaml file.

```
<?xml version="1.0" encoding="utf-8" ?>
<ContentPage xmlns="http://schemas.microsoft.com/dotnet/2021/maui"
             xmlns:x="http://schemas.microsoft.com/winfx/2009/xaml"
             x:Class="GridDesiredSizeBug.MainPage">

    <Grid BackgroundColor="#afa" VerticalOptions="Center" HorizontalOptions="Center">
       <Grid.RowDefinitions>
           <RowDefinition Height="*"/>
           <RowDefinition Height="auto"/>
       </Grid.RowDefinitions>
            <Label Grid.Row="0" Text="star desired 500" HeightRequest="500" BackgroundColor="#aaf"/>
            <Label Grid.Row="1" Text="auto" BackgroundColor="#faa"/>
    </Grid>

</ContentPage>
```

Start the app.
Change the screen orientation.

### Version with bug

Release Candidate 3 (current)

### Last version that worked well

Unknown/Other

### Affected platforms

Android

### Affected platform versions

Android 10

**Comment by AntonKosenkoDX** [comment]
Hello, we are developing [DevExpress .NET MAUI controls](https://www.devexpress.com/maui/) and experienced the same issue. This issue has a very high priority for us and we don't have a workaround. Thank you for your help.

**Comment by hartez** [comment]
The first part of this bug is fixed by #7116.

Please open a separate issue with your second bug so we can track it.
